pianoPlayer/main.js

```diff
@@ -0,0 +1,108 @@
+/* ISSUES & TODOs
+- add music!!
+- stop propagation / capture=true for when the keys are pressed (excluding the letters!)
+- mouseup event not caught when mouseout before
+- prevent key from overlapping the dark keys when it's being pressed
+- fix skeuomorphic design for upper keys
+*/
+
+// The keys and notes variables store the piano keys
+// plucks out the corresponding DOM elements and stores them as such
+const keys = ['c-key', 'd-key', 'e-key', 'f-key', 'g-key', 'a-key', 'b-key', 'high-c-key', 'c-sharp-key', 'd-sharp-key', 'f-sharp-key', 'g-sharp-key', 'a-sharp-key'];
+const notes = [];
+keys.forEach(function (key) {
+    notes.push(document.getElementById(key));
+})
```

Comment:
Ah! The first point of actual JS feedback.

Analyzing what this code is doing:
- defining an array
- iterating over another array
- for each of these iterations, add one corresponding element to the just-defined one

This is _exactly_ the time to use the method `.map`. `.map` returns a _transformed_ version of the original array that can accomplish exactly this.
```ts
const notes = keys.map(function (key) {
    return document.getElementById(key); 
});
```

Not only is this code terser & even a bit more legible, I'd say, but it also reduces a thing that might have a real term, but I call "code distance". It's the distance between where something is defined, modified, and used. I find that reducing code distance, and designing in such a way to incentivize that style of code, makes things more discoverable and easier to work with.

In this case, we save only one line, but it's entirely possible (and I see it all the time!) where people define the empty array alongside a lot of other definitions, and then like 10, 20 lines later actually load it up with values.

It's not a lot, but it adds to overhead and makes it harder to read.

Reply:
This is great yeah! This code was written by codecademy but I didn't think about touching it. Makes total sense tho! 😊